consensus/cp_mainvote.go

```diff
@@ -0,0 +1,60 @@
+package consensus
+
+import (
+	"github.com/pactus-project/pactus/crypto/hash"
+	"github.com/pactus-project/pactus/types/vote"
+)
+
+type cpMainVoteState struct {
+	*consensus
+}
+
+func (s *cpMainVoteState) enter() {
+}
+
+func (s *cpMainVoteState) decide() {
+	cpPreVotes := s.log.CPPreVoteVoteSet(s.round)
+	if cpPreVotes.HasTwoThirdOfTotalPower(s.cpRound) {
+		if cpPreVotes.HasQuorumVotesFor(s.cpRound, vote.CPValueOne) {
+			s.logger.Info("cp: quorum for pre-votes", "v", "1")
+
+			votes := cpPreVotes.BinaryVotes(s.cpRound, vote.CPValueOne)
+			cert := s.makeCertificate(votes)
+			just := &vote.JustMainVoteNoConflict{
+				QCert: cert,
+			}
+			s.signAddCPMainVote(hash.UndefHash, s.cpRound, vote.CPValueOne, just)
+			s.enterNewState(s.cpDecideState)
+		} else if cpPreVotes.HasQuorumVotesFor(s.cpRound, vote.CPValueZero) {
+			s.logger.Info("cp: quorum for pre-votes", "v", "0")
+
+			votes := cpPreVotes.BinaryVotes(s.cpRound, vote.CPValueZero)
+			cert := s.makeCertificate(votes)
+			just := &vote.JustMainVoteNoConflict{
+				QCert: cert,
+			}
+			s.signAddCPMainVote(*s.cpWeakValidity, s.cpRound, vote.CPValueZero, just)
+			s.enterNewState(s.cpDecideState)
+		} else {
+			s.logger.Info("cp: no-quorum for pre-votes", "v", "abstain")
+
+			vote0 := cpPreVotes.GetRandomVote(s.cpRound, vote.CPValueZero)
+			vote1 := cpPreVotes.GetRandomVote(s.cpRound, vote.CPValueOne)
+
+			just := &vote.JustMainVoteConflict{
+				Just0: vote0.CPJust(),
+				Just1: vote1.CPJust(),
+			}
+			s.signAddCPMainVote(*s.cpWeakValidity, s.cpRound, vote.CPValueAbstain, just)
+			s.enterNewState(s.cpDecideState)
+		}
+	}
+}
+
+func (s *cpMainVoteState) timeout(_ *ticker) {
+	// Ignore timeouts
```

Comment:
same question which mentioned upper

Reply:
In main-vote step we don't have timer.
These steps are so called full asynchronous steps.
It means there is no timer here and we should wait until we get 2/3+ votes. No matter how long.